prow/tide/search_test.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tide
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"testing"
+	"time"
+
+	githubql "github.com/shurcooL/githubv4"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/diff"
+)
+
+func TestSearch(t *testing.T) {
+	const q = "random search string"
+	now := time.Now()
+	earlier := now.Add(5 * time.Hour)
+	makePRs := func(numbers ...int) []PullRequest {
+		var prs []PullRequest
+		for _, n := range numbers {
+			prs = append(prs, PullRequest{Number: githubql.Int(n)})
+		}
+		return prs
+	}
+	makeQuery := func(more bool, cursor string, numbers ...int) searchQuery {
+		var sq searchQuery
+		sq.Search.PageInfo.HasNextPage = githubql.Boolean(more)
+		sq.Search.PageInfo.EndCursor = githubql.String(cursor)
+		for _, pr := range makePRs(numbers...) {
+			sq.Search.Nodes = append(sq.Search.Nodes, PRNode{pr})
+		}
+		return sq
+	}
+
+	cases := []struct {
+		name     string
+		start    time.Time
+		end      time.Time
+		q        string
+		cursors  []*githubql.String
+		sqs      []searchQuery
+		errs     []error
+		expected []PullRequest
+		err      bool
+	}{
+		{
+			name:    "single page works",
+			start:   earlier,
+			end:     now,
+			q:       datedQuery(q, earlier, now),
+			cursors: []*githubql.String{nil},
+			sqs: []searchQuery{
+				makeQuery(false, "", 1, 2),
+			},
+			errs:     []error{nil},
+			expected: makePRs(1, 2),
+		},
+		{
+			name:    "fail on first page",
+			start:   earlier,
+			end:     now,
+			q:       datedQuery(q, earlier, now),
+			cursors: []*githubql.String{nil},
+			sqs: []searchQuery{
+				{},
+			},
+			errs: []error{errors.New("injected error")},
+			err:  true,
+		},
+		{
+			name:    "set minimum start time",
```

Comment:
I don't think this actually validates that the start time is floored as expected.

Reply:
It does. We set the start time to `time.Time{}` (aka zero time). We validate this gets floored to `floor(time.Time{})` in the query

Reply:
Ah I see. This doesn't test that the behavior of `floor` is correct, it just tries to test that `floor` is called assuming the behavior is correct (which is reasonable given its simplicity).